Closes #9

peterbabic opened the pull request:
This PR fixes https://github.com/syntax-tree/unist-util-select/issues/9 

It is not a bug fix, but rather a code quality improvement. It's purpose is to stop JS bundler complaining about the detected circular dependency.

It does not include any new tests, because it would require testing for circular dependency, not sure what the right approach would be right now. The issue thread shows dependency graphs and a clear resolve of the cyclic dependency that existed before PR.

Note that possible state (context) pollution introduced by passing `any` down the road from the `index.js`

peterbabic committed:
move needsIndex to seprate file

Tries to remove cyclical dependency any.js -> psudo.js addressing
[#9](https://github.com/syntax-tree/unist-util-select/issues/9)

peterbabic committed:
remove ciruclar dependency passing any via state

peterbabic committed:
move needsIndex to any

wooorm commented on index.js:
```diff
@@ -8,7 +8,7 @@ var any = require('./lib/any')
 
 function matches(selector, node) {
-  return Boolean(any(parse(selector), node, {one: true, shallow: true})[0])
+  return Boolean(any(parse(selector), node, {one: true, shallow: true, any})[0])
```
```suggestion
  return Boolean(any(parse(selector), node, {one: true, shallow: true, any: any})[0])
```

wooorm replied:
and shouldn’t this also be added to the bottom ones?

peterbabic replied:
I am not sure. To me it looks like only the two functions are affected, already updated in the PR. How do we check apart from passing tests that it does not need to be done?

wooorm replied:
`select` and `selectAll` also should support pseudo elements. Right now only `matches` has `any`. But this is an assumption. It seems there are no tests for pseudos in `select{All}`?

wooorm replied:
this one is I think untested and a bug

peterbabic replied:
What about https://github.com/syntax-tree/unist-util-select/blob/2269b1c4fc8dc8c824d2757612c7c398ae0da80a/test/select.js#L228 and https://github.com/syntax-tree/unist-util-select/blob/2269b1c4fc8dc8c824d2757612c7c398ae0da80a/test/select-all.js#L244

They check for pseudo element `:first-child` using `select` and `selectAll` method if I am not mistaken

peterbabic replied:
Susequently, I believe the pseudo-elements tests for select and selectAll go all the way down to L696 or L730 respectively, covering quite a lot. Not a single one tripped. Am I being mistaken here?

wooorm commented on lib/any.js:
```diff
@@ -2,8 +2,20 @@
 
 module.exports = match
 
+match.needsIndex = [
```
```suggestion
var needsIndex = [
```

I don’t think this needs to be exported

wooorm commented on lib/any.js:
```diff
@@ -75,7 +88,7 @@ function configure(query, state) {
 
   while (++index < pseudos.length) {
-    if (pseudo.needsIndex.indexOf(pseudos[index].name) > -1) {
+    if (match.needsIndex.indexOf(pseudos[index].name) > -1) {
```
```suggestion
    if (needsIndex.indexOf(pseudos[index].name) > -1) {
```

wooorm reviewed (commented):
thanks for taking a stab at this!

peterbabic committed:
make needsIndex private

Co-authored-by: Titus <tituswormer@gmail.com>

peterbabic committed:
make needsIndex private

Co-authored-by: Titus <tituswormer@gmail.com>

wooorm commented on lib/pseudo.js:
```diff
@@ -5,23 +5,9 @@ module.exports = match
 
 var is = convert()
 
-match.needsIndex = [
```
this change can be reverted

wooorm commented on lib/any.js:
```diff
@@ -75,7 +88,7 @@ function configure(query, state) {
 
   while (++index < pseudos.length) {
-    if (pseudo.needsIndex.indexOf(pseudos[index].name) > -1) {
+    if (needsIndex.indexOf(pseudos[index].name) > -1) {
```
```suggestion
    if (pseudo.needsIndex.indexOf(pseudos[index].name) > -1) {
```

this change can be reverted

wooorm commented on lib/any.js:
```diff
@@ -2,8 +2,20 @@
+]
+
 var zwitch = require('zwitch')
-var pseudo = require('./pseudo')
```
this change can be reverted

peterbabic replied:
I am getting quite lost here :sad: 

What could be reverted, please?

wooorm replied:
We were discussing two approaches.
1. place `needsIndex` in `any`, remove it from `pseudo`, which doesn’t work.
2. Pass `any` down

As 1 didn’t work, it can be reverted

peterbabic replied:
Pardon me, but I am not seeing why it doesn't work. Could you please point me to the right direction?

wooorm replied:
You said [here in the issue](https://github.com/syntax-tree/unist-util-select/issues/9#issuecomment-744621815) that changing `needsIndex` had no effect

peterbabic replied:
I have made another commit https://github.com/syntax-tree/unist-util-select/pull/10/commits/89f87356ac213f685d5697446632c073f9e64a78 which I believe places `needsIndex` into even better place, by still not tripping any tests and with the circular dependency resolved.

If there is need to revert something, it has to be by your reasoning as a reviewer, otherwise I suspect I wasn't clear somewhere in the issue and your interpretation of what I have stated is different from what I have meant.

Please, provide way of moving this PR further, discarding the "this change could be reverted" suggestions in the process, if possible.

peterbabic replied:
Also, I might start getting what do you mean. To make things clear further, BOTH "approaches" need to be applied, not just the second one (with reverting the first one) to remove the cyclic dependency. 

It is true that I said the first one does not work, meaning it reduced the dependency from direct (A -> B -> A) to indirect (A -> B -> C -> A, indirect cyclic dependencies tend to be easier to resolve) which was then resolved by the approach 2.

peterbabic replied:
Or in even different words, there are TWO cyclic dependencies to begin with, the first approach removes the direct one (yellow), the second approach removes the indirect one (red)

![2020-12-15_11-50](https://user-images.githubusercontent.com/3952052/102205679-cdfb1700-3ecb-11eb-9227-74dccdc03c82.png)

wooorm commented on lib/any.js:
```diff
@@ -2,8 +2,20 @@
 
 module.exports = match
 
+var needsIndex = [
+  'first-child',
+  'first-of-type',
+  'last-child',
+  'last-of-type',
+  'nth-child',
+  'nth-last-child',
+  'nth-of-type',
+  'nth-last-of-type',
+  'only-child',
+  'only-of-type'
+]
+
```
```suggestion
```

this change can be reverted

wooorm committed:
Update index.js

wooorm reviewed (commented):
The recursiveness seems to only be needed in [`matches`](https://github.com/syntax-tree/unist-util-select/blob/2269b1c4fc8dc8c824d2757612c7c398ae0da80a/lib/pseudo.js#L61) (`x:matches(:link)`) and [`hasSelector`](https://github.com/syntax-tree/unist-util-select/blob/2269b1c4fc8dc8c824d2757612c7c398ae0da80a/lib/pseudo.js#L158) (`x:has(:link)`). I’m assuming those two aren’t tested with pseudos in the `select` or `selectAll` cases

peterbabic committed:
move needsIndex into local scope

Configure function is the sole consumer of the needsIndex array, it
shoould then be in it's local scope. This change further improves code
quiality.

wooorm commented:
I don’t see why the changes previously applied for approach 1 are needed. Moving `any` to something that is passed around removes the recursive dep.

Could you make the changes in this PR as small as possible? Only the change that `any` is passed around, not the `needsIndex` changes.

And could you pass `any: any`  in `index.js` to the two other functions too?

peterbabic commented:
Or in even different words, there are TWO cyclic dependencies to begin with, the first approach removes the direct one (yellow), the second approach removes the indirect one (red)

![2020-12-15_11-50](https://user-images.githubusercontent.com/3952052/102205679-cdfb1700-3ecb-11eb-9227-74dccdc03c82.png)

wooorm commented:
By passing any down to all files, that is removed though?

wooorm commented:
Hold on, let me try and do it myself. I’m sorry that I can’t seem to make my suggestion clear.

peterbabic commented:
I am sorry, you are right. Just by applying the approach 2 the cyclic dependency is resolved. 

However by passing `any: any` into contextx into select and selectAll does not seem to affect anything

wooorm commented:
I described `select` and `selectAll` here: https://github.com/syntax-tree/unist-util-select/pull/10#discussion_r542764202

peterbabic commented:
Well, to my best efforts:
-  Approach 2, minimal code needed for tests passing and cyclic dependency removed
https://github.com/peterbabic/unist-util-select/commit/375cdfe1672aaec2791e308c95f0a46942a6f158

- Approach 2 + passing any also into select and selectAll, not just into matches
https://github.com/peterbabic/unist-util-select/commit/e455227c42aab163293e07901b726698b9ca0906

These two could be squashed if needed, but TBH I am not quite there yet about the https://github.com/syntax-tree/unist-util-select/pull/10#discussion_r542764202

wooorm commented:
I’m already working on that approach, thanks, will push soon (but install graphviz for marge broke my computer 🥺)
